src/machine/machine_avr.go

```diff
@@ -271,3 +272,125 @@ func (i2c I2C) ReadByte() byte {
 
 	return byte(*avr.TWDR)
 }
+
+// UART
+
+type UARTConfig struct {
+	Baudrate uint32
+}
+
+type UART struct {
+}
+
+var (
+	// UART0 is the hardware serial port on the AVR.
+	UART0 = &UART{}
+)
+
+// Configure the UART on the AVR.
+func (uart UART) Configure(config UARTConfig) {
+	// default to 115200
+	if config.Baudrate == 0 {
+		config.Baudrate = 115200
+	}
+
+	// Set baud rate based on prescale formula:
+	// BAUD_PRESCALE = (((F_CPU / (BAUDRATE * 16UL))) - 1)
+	ps := (CPU_FREQUENCY / (config.Baudrate * 16)) - 1
+	*avr.UBRR0H = avr.RegValue(ps >> 8)
+	*avr.UBRR0L = avr.RegValue(ps)
+
+	// enable RX interrupt
+	*avr.UCSR0B |= avr.UCSR0B_RXCIE0
+}
+
+// Close the UART on the AVR.
+func (uart UART) Close() error {
+	return nil
+}
+
+// Read from the RX buffer.
+func (uart UART) Read(data []byte) (n int, err error) {
+	// check if RX buffer is empty
+	size := uart.Buffered()
+	if size == 0 {
+		return 0, nil
+	}
+
+	// only read number of bytes used from buffer
+	for i := 0; i < size; i++ {
+		v, _ := uart.ReadByte()
+		data[i] = byte(v)
+	}
+
+	return len(data), nil
+}
+
+// Write data to the UART.
+func (uart UART) Write(data []byte) (n int, err error) {
+	for _, v := range data {
+		uart.WriteByte(v)
+	}
+	return len(data), nil
+}
+
+// ReadByte reads a single byte from the RX buffer.
+func (uart UART) ReadByte() (byte, error) {
+	// check if RX buffer is empty
+	if uart.Buffered() == 0 {
+		return 0, errors.New("Buffer empty")
+	}
+
+	return bufferGet(), nil
+}
+
+// WriteByte writes a byte of data to the UART.
+func (uart UART) WriteByte(c byte) error {
+	// Wait until UART buffer is not busy.
+	for (*avr.UCSR0A & avr.UCSR0A_UDRE0) == 0 {
+	}
+	*avr.UDR0 = avr.RegValue(c) // send char
+	return nil
+}
+
+// Buffered returns the number of bytes current stored in the RX buffer.
+func (uart UART) Buffered() int {
+	return int(bufferUsed())
+}
+
+type __volatile byte
```

Comment:
I just pushed a commit that makes marking variables as volatile a bit more flexible, which means this code breaks.
Now, you can set a named type as volatile by adding the `//go:volatile` pragma, for example:

```go
//go:volatile
type volatileByte byte

var head volatileByte
```

Another example: https://github.com/aykevl/tinygo/blob/master/src/runtime/runtime_nrf.go#L85

Reply:
Note: this also means the register descriptors need to be regenerated (`make gen-device-avr` etc.).